### ISSUE TITLE
Question: State of link (sender | receiver) after it gets detached or the connection gets closed.

### ISSUE THREAD

**Issue by amarzavery** [amqp/rhea#104]
I have a simple test for Event Hub where I create 1 sender link on 1 connection and 1 receiver link on another connection. _(Note: Every link has its dedicated session)_.

Test: Sends one message with message_id and verifies that the receiver received that message (Verifies that there is no message loss). This sending and receiving is done once every 31 minutes. EventHubs detaches a link (sender | receiver) if it is inactive for more than 30 minutes. Test also ensures that the sdk revives the link from such errors.

The test was running fine for ~10 hours. After that due to some issues on the server side (because it was moving nodes internally) different type of connection issues started to happen. Sometimes the server sends empty detach frames for the link, sometimes the detach frame has an error and sometimes the connection gets disconnected (I see errors like `This socket has been ended by the other party`).

Now I have reconnect logic for 2 scenarios.
1. Connection is fine. The link got detached due to inactivity or some other error and gets a detach frame. I have a listener for `receiver|sender_close` where I detect if the condition seems favorable for re-establishing the link then I do so.
2. Connection gets the `disconnected` event. At that time I look in my internal map for the sender and receiver links that were present and try to re-establish them. (Note: The sender/receiver link, try to do cbs authentication as a part of their initialization. The cbs auth is the central piece that makes sure to bring the connection up if it is down, or use the connection if it is already up. Thus reviving the first link will revive the connection and the links thereafter will use that connection).

Question.
1. In the first scenario, I saw that every time I get a detach frame and try to re-establish the link, I see that the channel number increases. In my reconnect method, I am not closing the existing link with it's session. Should I **close** the link and it's underlying session? OR Should I **remove** the link and it's underlying session? The reason for asking this is seen in the following logs, where I see a bunch of end frames for both the sender and receiver links.

I was surprised to see so many session ends. Ideally I should have only 2 of them. One for sender link's session and one for receiver link's session. This makes me think that I need to do something explicitly while the links have been detached. Can you provide some guidance over here?

>Following statements were logged so many times because the "session_close" event was called that many times
```bash
2018-07-19T17:58:44.857Z azure:amqp-common:retry Retry attempt number: 1
2018-07-19T17:58:44.858Z azure:event-hubs:sender [connection-2] Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' is open? -> true
2018-07-19T17:58:44.858Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.858Z rhea:frames [connection-2]:2 <- end#17 {} 
2018-07-19T17:58:44.858Z rhea:events Session got event: session_close
2018-07-19T17:58:44.858Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.858Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.858Z rhea:frames [connection-2]:3 <- end#17 {} 
2018-07-19T17:58:44.858Z rhea:events Session got event: session_close
2018-07-19T17:58:44.858Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.858Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.858Z rhea:frames [connection-2]:4 <- end#17 {} 
2018-07-19T17:58:44.858Z rhea:events Session got event: session_close
2018-07-19T17:58:44.858Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.858Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.858Z rhea:frames [connection-2]:5 <- end#17 {} 
2018-07-19T17:58:44.858Z rhea:events Session got event: session_close
2018-07-19T17:58:44.858Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.858Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.858Z rhea:frames [connection-2]:6 <- end#17 {} 
2018-07-19T17:58:44.858Z rhea:events Session got event: session_close
2018-07-19T17:58:44.858Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.859Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.859Z rhea:frames [connection-2]:7 <- end#17 {} 
2018-07-19T17:58:44.859Z rhea:events Session got event: session_close
2018-07-19T17:58:44.859Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.859Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.859Z rhea:frames [connection-2]:8 <- end#17 {} 
2018-07-19T17:58:44.859Z rhea:events Session got event: session_close
2018-07-19T17:58:44.859Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.859Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.859Z rhea:frames [connection-2]:9 <- end#17 {} 
2018-07-19T17:58:44.860Z rhea:events Session got event: session_close
2018-07-19T17:58:44.860Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.860Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.860Z rhea:frames [connection-2]:10 <- end#17 {} 
2018-07-19T17:58:44.860Z rhea:events Session got event: session_close
2018-07-19T17:58:44.860Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.860Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.860Z rhea:frames [connection-2]:11 <- end#17 {} 
2018-07-19T17:58:44.860Z rhea:events Session got event: session_close
2018-07-19T17:58:44.860Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.860Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.860Z rhea:frames [connection-2]:12 <- end#17 {} 
2018-07-19T17:58:44.860Z rhea:events Session got event: session_close
2018-07-19T17:58:44.860Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.860Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.860Z rhea:frames [connection-2]:13 <- end#17 {} 
2018-07-19T17:58:44.860Z rhea:events Session got event: session_close
2018-07-19T17:58:44.860Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.860Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.860Z rhea:frames [connection-2]:14 <- end#17 {} 
2018-07-19T17:58:44.860Z rhea:events Session got event: session_close
2018-07-19T17:58:44.860Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.860Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.861Z rhea:frames [connection-2]:15 <- end#17 {} 
2018-07-19T17:58:44.861Z rhea:events Session got event: session_close
2018-07-19T17:58:44.861Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.861Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.861Z rhea:frames [connection-2]:16 <- end#17 {} 
2018-07-19T17:58:44.861Z rhea:events Session got event: session_close
2018-07-19T17:58:44.861Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.861Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.861Z rhea:frames [connection-2]:17 <- end#17 {} 
2018-07-19T17:58:44.861Z rhea:events Session got event: session_close
2018-07-19T17:58:44.861Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.861Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.861Z rhea:frames [connection-2]:18 <- end#17 {} 
2018-07-19T17:58:44.861Z rhea:events Session got event: session_close
2018-07-19T17:58:44.861Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.861Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.861Z rhea:frames [connection-2]:19 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:20 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:21 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:22 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:23 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:24 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:25 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z rhea:events Connection got event: session_close
2018-07-19T17:58:44.863Z rhea:events Container got event: session_close
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:26 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:27 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z rhea:events Connection got event: session_close
2018-07-19T17:58:44.863Z rhea:events Container got event: session_close
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:28 <- end#17 {} 
2018-07-19T17:58:44.863Z rhea:events Session got event: session_close
2018-07-19T17:58:44.863Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.863Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.863Z rhea:frames [connection-2]:29 <- end#17 {} 
2018-07-19T17:58:44.864Z rhea:events Session got event: session_close
2018-07-19T17:58:44.864Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.864Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.864Z rhea:frames [connection-2]:30 <- end#17 {} 
2018-07-19T17:58:44.864Z rhea:events Session got event: session_close
2018-07-19T17:58:44.864Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.864Z rhea:io [connection-2] got frame of size 17
2018-07-19T17:58:44.864Z rhea:frames [connection-2]:31 <- detach#16 {"closed":true} 
2018-07-19T17:58:44.865Z rhea:events Link got event: sender_close
2018-07-19T17:58:44.865Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.865Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.865Z rhea:frames [connection-2]:31 <- end#17 {} 
2018-07-19T17:58:44.865Z rhea:events Session got event: session_close
2018-07-19T17:58:44.865Z azure:event-hubs:sender [connection-2] close() method of Sender '2fc17808-d056-4683-8873-e52fdabc79c1' with address 'ehnode3/Partitions/0' was not called. There was no accompanying error as well. This is a candidate for re-establishing the sender link.
2018-07-19T17:58:44.865Z rhea:io [connection-2] got frame of size 15
2018-07-19T17:58:44.865Z rhea:frames [connection-2]:0 <- close#18 {} 
2018-07-19T17:58:44.865Z rhea:events Connection got event: connection_close
2018-07-19T17:58:44.865Z rhea:events Container got event: connection_close
2018-07-19T17:58:44.866Z rhea:frames [connection-2]:0 -> end#17 {} 
2018-07-19T17:58:44.866Z rhea:frames [connection-2]:1 -> end#17 {} 
2018-07-19T17:58:44.866Z rhea:frames [connection-2]:2 -> end#17 {} 
2018-07-19T17:58:44.866Z rhea:frames [connection-2]:3 -> end#17 {} 
2018-07-19T17:58:44.866Z rhea:frames [connection-2]:4 -> end#17 {} 
2018-07-19T17:58:44.866Z rhea:frames [connection-2]:5 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:6 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:7 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:8 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:9 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:10 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:11 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:12 -> end#17 {} 
2018-07-19T17:58:44.867Z rhea:frames [connection-2]:13 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:14 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:15 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:16 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:17 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:18 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:19 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:20 -> end#17 {} 
2018-07-19T17:58:44.868Z rhea:frames [connection-2]:21 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:22 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:23 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:24 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:25 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:26 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:27 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:28 -> end#17 {} 
2018-07-19T17:58:44.869Z rhea:frames [connection-2]:29 -> end#17 {} 
2018-07-19T17:58:44.870Z rhea:frames [connection-2]:30 -> end#17 {} 
2018-07-19T17:58:44.870Z rhea:frames [connection-2]:31 -> end#17 {} 
2018-07-19T17:58:44.870Z rhea:frames [connection-2]:0 -> close#18 {} 
```

2. In the second scenario, after the `"disconnected"` event is received,
- what will happen if we execute `sender.close()` or `receiver.close()`?
- Should we call remove() instead of close() (since the connection is dead, there is no point, keeping them in the local_channel_map or remote_channel_map)?

**Comment by grs** [comment]
Are you creating sessions? By default the library uses a single session per connection.

**Comment by amarzavery** [comment]
Should I close or remove the session? If I call session.close() after the disconnected event is received (i.e the connection is down) then what will be the effect on the internal map? Will the session be deleted or still be present?

**Comment by grs** [comment]
If connected, you should close it, wait for the peer to also close it, then remove it. (We can make that simpler, raise an issue for it if you like).

If disconnected, you can just remove the session.

**Comment by tony-gutierrez** [comment]
@amarzavery I think we are working on the same thing. Clean reconnects. When rhea reconnects me I get all kinds of errors that $cbs sender already exists and errors that the receiver is unauthorized before my code gets to the $cbs token send. 

I have tried closing the receivers on disconnect but it seems that the remote remembers them.

**Comment by grs** [comment]
Fyi, I am going on holiday for a few days (25th-30th July). I had hoped to get to this before that, but looks like that will not happen today. If it is urgent, I suggest you workaround it in the application by closing/removing as needed. I'll get a fix into the library to handle this on my return.

**Comment by amarzavery** [comment]
@grs - I tried to close/remove the session in the application and was getting different type of issues. It would be great if you can provide a fix for this in rhea.

**Comment by grs** [comment]
See https://github.com/amqp/rhea/pull/113 which I believe should fix this.

**Comment by amarzavery** [comment]
fixed by #113
